crates/ide_completion/src/context.rs

```diff
@@ -290,23 +292,114 @@ impl<'a> CompletionContext<'a> {
         file_with_fake_ident: SyntaxNode,
         offset: TextSize,
     ) {
-        // FIXME: this is wrong in at least two cases:
-        //  * when there's no token `foo($0)`
-        //  * when there is a token, but it happens to have type of it's own
-        self.expected_type = self
-            .token
-            .ancestors()
-            .find_map(|node| {
-                let ty = match_ast! {
+        let expected = {
+            let mut node = self.token.parent();
+            loop {
+                let ret = match_ast! {
                     match node {
-                        ast::Pat(it) => self.sema.type_of_pat(&it),
-                        ast::Expr(it) => self.sema.type_of_expr(&it),
-                        _ => return None,
+                        ast::LetStmt(it) => {
+                            cov_mark::hit!(expected_type_let_with_leading_char);
+                            cov_mark::hit!(expected_type_let_without_leading_char);
+                            let ty = it.pat()
+                                .map(|pat| self.sema.type_of_pat(&pat))
+                                .flatten();
+
+                            (ty, None)
+                        },
+                        ast::ArgList(it) => {
+                            cov_mark::hit!(expected_type_fn_param_with_leading_char);
+                            cov_mark::hit!(expected_type_fn_param_without_leading_char);
+                            ActiveParameter::at_token(
+                                &self.sema,
+                                self.token.clone(),
+                            ).map(|ap| (Some(ap.ty), Some(ap.name)))
+                            .unwrap_or((None, None))
+                        },
+                        ast::RecordExprFieldList(it) => {
+                            cov_mark::hit!(expected_type_struct_field_without_leading_char);
+                            // FIXME: This doesn't work in the situation we'd
+                            // like to use it, when the user has not yet typed
+                            // any part of the struct field expression:
+                            //
+                            // ```
+                            // struct Foo { a: u32 }
+                            // fn foo() { Foo { a: $0 } }
+                            // ```
+                            //
+                            // The trick we are using here is in this case the
+                            // users cursor will be associated with a whitespace
+                            // token which is a sibling to the ast::RecordExprField.
+                            //
+                            // The failure happens because resolve_record_field expects
+                            // the `expr` of that field to be `Some`, which it is not.
+                            //
+                            // Is there some way to directly look up the type of the
+                            // NameRef `a`?
```

Comment:
You might be looking for https://github.com/rust-analyzer/rust-analyzer/blob/0c2f570151749d14e351f30fa234693b36bc4e72/crates/syntax/src/ast/node_ext.rs#L275-L282

Due to field shorthand, the AST situation here is non-trivial. 

also, a bunch of map flattens can be converted to and_thens

Reply:
> You might be looking for

I don't *think* this is what I need. I already have some code which gets the `ast::RecordExprField`, but then `Semantics::resolve_record_field` returns `None` because in this case my record field has no `expr`.

This is when the source code looks like:

```rust
struct A { a: u32 }

fn foo() {
    A { a: $0 };
}
```

I'm fairly certain the existing implementation (before this PR) has the same problem here, so I'd be happy to push this off to a future PR if the solution is complex.

Reply:
yeah, plus one to postponing this!

Reply:
This has been resolved after rebasing since #8025 merged :tada: 

I've updated the tests and removed the FIXME comments in both places!